warehouse/schema/schema.go

```diff
@@ -192,23 +221,9 @@ func (sh *schema) isIDResolutionEnabled() bool {
 }
 
 func (sh *schema) TableSchemaDiff(ctx context.Context, tableName string, tableSchema model.TableSchema) (whutils.TableSchemaDiff, error) {
-	schema, err := sh.getSchema(ctx)
-	if err != nil {
-		return whutils.TableSchemaDiff{}, fmt.Errorf("getting schema: %w", err)
-	}
-	return tableSchemaDiff(tableName, schema, tableSchema), nil
-}
-
-func (sh *schema) fetchSchemaFromWarehouse(ctx context.Context) (model.Schema, error) {
-	start := sh.now()
-	warehouseSchema, err := sh.fetchSchemaRepo.FetchSchema(ctx)
-	if err != nil {
-		return nil, fmt.Errorf("fetching schema: %w", err)
-	}
-	duration := math.Round((sh.now().Sub(start).Minutes() * 1000)) / 1000
-	sh.log.Infon("Fetched schema from warehouse", obskit.DestinationID(sh.warehouse.Destination.ID), obskit.Namespace(sh.warehouse.Type), logger.NewFloatField("timeTakenInMinutes", duration))
-	removeDeprecatedColumns(warehouseSchema, sh.warehouse, sh.log)
-	return warehouseSchema, sh.saveSchema(ctx, warehouseSchema)
+	sh.cachedSchemaMu.RLock()
+	defer sh.cachedSchemaMu.RUnlock()
+	return tableSchemaDiff(tableName, sh.cachedSchema, tableSchema), nil
 }
 
 func (sh *schema) saveSchema(ctx context.Context, newSchema model.Schema) error {
```

Comment:
minor: This we can discuss later while we start testing this fix. Should this func be thread safe instead of usage to be thread safe? Later one, chances of missing to consider this would be high.

Reply:
> Should this func be thread safe

This function only executes a Postgres query and doesn't perform any other operations, so there's no need to use a lock. In the case of concurrent calls from multiple goroutines, the final DB value will be that of the last write.

Reply:
1. In our usage, we are taking lock to update DB and memory cache both which is right.
 2. `the final DB value will be that of the last write` -- I am still doubtful about this. Say three goroutines are planning to write value A, B, C. Last write can be any of these? Unless we are using it thread safe as in point 1.

Reply:
By last write, I meant whoever finished last and not in the order which they invoked this function

> Say three goroutines are planning to write value A, B, C. Last write can be any of these?

Yes